Enhancement: optimize bulk custom field operations for performance

cdjk opened the pull request:
## Proposed change

Previously bulk-editing a custom field would result in a sql update being issued in a for loop. This was slow - on the order of seconds per document.

This needs separate operations for bulk_create + bulk_update in order to support mysql/mariadb - django supports bulk_create(update_conflicts=True) for postgres and sqlite.

(all times in seconds)

|  | time - before change | time - after change |
| -- | --- | --- |
| sqlite - 10 documents | 26 | .14 |
| sqlite - 100 documents | 124 | .15 |
| postgres - 10 documents | 59 | .15 |
| postgres - 100 documents |199 | .21 |
| mariadb - 10 documents | 19 | .14 |
| mariadb - 100 documents | 209 | .16 |

I used AI to write:
- the performance testing code
- some of the doclink reflection bulk updates.

## Type of change

- [ ] Bug fix: non-breaking change which fixes an issue.
- [ ] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [x] Other. Please explain: performance improvement - no functional change

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for breaking changes & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] In the description of the PR above I have disclosed the use of AI tools in the coding of this PR.

cdjk commented:
It's real data - I used the document exporter on my paperless install, which has about 2.5k files, then the document importer to load. This matches the performance I noticed, which is why I wanted to fix it.

I'm running on a t3.large. The performance is about the same with postgres on RDS or in docker. I tried it on my mac (m4 pro) with just postgres, and got:

| number of documents | time bulk edit a field |
| -- | --- |
| 10 | 5 |
| 20 | 9 |
| 50 | 20 |
| 100 | 40 |
| 200 | 68 |  

There's shell script that calls starts the container using the official docker compose files (with an override to specify my locally built image), and a manage.py command that bulk edits a field. Happy to share those, or check them in if there's a good spot. I wrote the shell script, AI wrote the manage.py command (but it's simple).

I'm happy to revert the doclink reflection code. Editing 100 or so custom fields at a time with a document links seems contrived, but bulk setting other custom fields, especially when loading up a bunch of things, feels fairly common.

If there's something obvious I'm missing in my setup that would speed things up I'm happy to do that instead.

shamoon commented:
Cool thanks. It’s interesting, not my experience but I’ll have to play with it.

But I am sure there is lots of room to improve and we’d be happy to realize those gains. I suppose I’d be curious if the first part (not the reflect stuff) is enough to see the significant reduction. I do wonder if that might strike a better balance of performance vs “risk”. We’ve come to accept paperless is just kinda complicated, sometimes even small changes have surprising consequences…

cdjk committed:
Optimize bulk custom field operations for performance
- Replace per-document update_or_create loop with bulk operations
  (bulk_create + bulk_update) for better performance. Can't use
  bulk_create(update_conflicts=True) because that only works on
  PostgreSQL/SQLite.
- Do not optimize doclinks for now - too complex

- Add warning log when attempting to add a non-existent custom field.

cdjk commented:
Closing, see #11558
